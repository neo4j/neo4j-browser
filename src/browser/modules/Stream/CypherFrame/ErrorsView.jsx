/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from 'react'
import { withBus } from 'react-suber'
import Ellipsis from 'browser-components/Ellipsis'
import {
  ExclamationTriangleIcon,
  PlayIcon
} from 'browser-components/icons/Icons'
import { deepEquals } from 'services/utils'
import Render from 'browser-components/Render'
import {
  executeCommand,
  listDbsCommand
} from 'shared/modules/commands/commandsDuck'
import { listAvailableProcedures } from 'shared/modules/cypher/procedureFactory'
import {
  isUnknownProcedureError,
  isNoDbAccessError
} from 'services/cypherErrorsHelper'
import { errorMessageFormater } from './../errorMessageFormater'

import {
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledErrorH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledLink,
  StyledLinkContainer,
  StyledHelpFrame,
  ErrorText
} from '../styled'

export class ErrorsView extends Component {
  shouldComponentUpdate(props, state) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const { result: error, bus } = this.props
    if (!error || !error.code) {
      return null
    }
    const fullError = errorMessageFormater(null, error.message)

    return (
      <StyledHelpFrame>
        <StyledHelpContent>
          <StyledHelpDescription>
            <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
            <StyledErrorH4>{error.code}</StyledErrorH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledPreformattedArea>{fullError.message}</StyledPreformattedArea>
          </StyledDiv>
          <Render if={isUnknownProcedureError(error)}>
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, listAvailableProcedures)}
              >
                <PlayIcon />
                &nbsp;List available procedures
              </StyledLink>
            </StyledLinkContainer>
          </Render>
          <Render if={isNoDbAccessError(error)}>
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, `:${listDbsCommand}`)}
              >
                <PlayIcon />
                &nbsp;List available databases
              </StyledLink>
            </StyledLinkContainer>
          </Render>
        </StyledHelpContent>
      </StyledHelpFrame>
    )
  }
}

const onItemClick = (bus, statement) => {
  const action = executeCommand(statement)
  bus.send(action.type, action)
}

export const ErrorsViewBus = withBus(ErrorsView)

export class ErrorsStatusbar extends Component {
  shouldComponentUpdate(props, state) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const error = this.props.result
    if (!error || (!error.code && !error.message)) return null
    const fullError = errorMessageFormater(error.code, error.message)

    return (
      <Ellipsis>
        <ErrorText title={fullError.title}>
          <ExclamationTriangleIcon /> {fullError.message}
        </ErrorText>
      </Ellipsis>
    )
  }
}
