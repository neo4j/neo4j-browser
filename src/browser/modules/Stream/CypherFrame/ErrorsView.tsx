/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { Component } from 'react'
import { withBus } from 'react-suber'
import Ellipsis from 'browser-components/Ellipsis'
import {
  ExclamationTriangleIcon,
  PlayIcon
} from 'browser-components/icons/Icons'
import { deepEquals } from 'services/utils'

import {
  commandSources,
  executeCommand,
  listDbsCommand
} from 'shared/modules/commands/commandsDuck'
import { listAvailableProcedures } from 'shared/modules/cypher/procedureFactory'
import {
  isUnknownProcedureError,
  isNoDbAccessError,
  isPeriodicCommitError
} from 'services/cypherErrorsHelper'
import { errorMessageFormater } from './../errorMessageFormater'

import {
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledErrorH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledLink,
  StyledLinkContainer,
  StyledHelpFrame,
  ErrorText
} from '../styled'

export class ErrorsView extends Component<any> {
  shouldComponentUpdate(props: any) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const { result: error, bus } = this.props
    if (!error || !error.code) {
      return null
    }
    const fullError = errorMessageFormater(null, error.message)

    return (
      <StyledHelpFrame>
        <StyledHelpContent>
          <StyledHelpDescription>
            <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
            <StyledErrorH4>{error.code}</StyledErrorH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledPreformattedArea>{fullError.message}</StyledPreformattedArea>
          </StyledDiv>
          {isUnknownProcedureError(error) && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, listAvailableProcedures)}
              >
                <PlayIcon />
                &nbsp;List available procedures
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isNoDbAccessError(error) && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() => onItemClick(bus, `:${listDbsCommand}`)}
              >
                <PlayIcon />
                &nbsp;List available databases
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isPeriodicCommitError(error) && (
            <StyledLinkContainer>
              <StyledLink onClick={() => onItemClick(bus, `:help auto`)}>
                <PlayIcon />
                &nbsp;Info on the <code>:auto</code> command
              </StyledLink>
              &nbsp;(auto-committing transactions)
            </StyledLinkContainer>
          )}
        </StyledHelpContent>
      </StyledHelpFrame>
    )
  }
}

const onItemClick = (bus: any, statement: any) => {
  const action = executeCommand(statement, { source: commandSources.button })
  bus.send(action.type, action)
}

export const ErrorsViewBus = withBus(ErrorsView)

export class ErrorsStatusbar extends Component<any> {
  shouldComponentUpdate(props: any) {
    return !deepEquals(props.result, this.props.result)
  }

  render() {
    const error = this.props.result
    if (!error || (!error.code && !error.message)) return null
    const fullError = errorMessageFormater(error.code, error.message)

    return (
      <Ellipsis>
        <ErrorText title={fullError.title}>
          <ExclamationTriangleIcon /> {fullError.message}
        </ErrorText>
      </Ellipsis>
    )
  }
}
