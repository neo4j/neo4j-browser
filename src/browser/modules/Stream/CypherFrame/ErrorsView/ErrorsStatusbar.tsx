/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react'

import { deepEquals, ExclamationTriangleIcon } from 'nxComps/common'

import Ellipsis from 'browser-components/Ellipsis'
import { errorMessageFormater } from 'project-root/src/browser/modules/Stream/errorMessageFormater'
import { ErrorText } from 'project-root/src/browser/modules/Stream/styled'
import { BrowserRequestResult } from 'project-root/src/shared/modules/requests/requestsDuck'
import { BrowserError } from 'services/exceptions'

type ErrorsStatusBarProps = {
  result: BrowserRequestResult
}
export class ErrorsStatusbar extends Component<ErrorsStatusBarProps> {
  shouldComponentUpdate(props: ErrorsStatusBarProps): boolean {
    return !deepEquals(props.result, this.props.result)
  }

  render(): null | JSX.Element {
    const error = this.props.result as BrowserError
    if (!error || (!error.code && !error.message)) return null
    const fullError = errorMessageFormater(error.code, error.message)

    return (
      <Ellipsis>
        <ErrorText title={fullError.title}>
          <ExclamationTriangleIcon /> {fullError.message}
        </ErrorText>
      </Ellipsis>
    )
  }
}
