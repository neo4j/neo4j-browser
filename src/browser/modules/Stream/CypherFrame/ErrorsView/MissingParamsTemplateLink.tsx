/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React from 'react'

import {
  StyledMissingParamsTemplateLink,
  StyledParamsTemplateClickableArea,
  StyledSpecifyParamsText
} from 'project-root/src/browser/modules/Stream/styled'
import { stringModifier } from 'services/bolt/cypherTypesFormatting'
import { BrowserError } from 'services/exceptions'
import { stringifyMod } from 'services/utils'

type MissingParamsTemplateLinkProps = {
  error: BrowserError
  params: Record<string, unknown>
  setEditorContent: (cmd: string) => void
  onTemplateHelpMessageClick: () => void
}
export const MissingParamsTemplateLink = ({
  setEditorContent,
  params,
  error,
  onTemplateHelpMessageClick
}: MissingParamsTemplateLinkProps): JSX.Element => {
  const handleTemplateHelpMessageClick = (): void => {
    onTemplateHelpMessageClick()
    const missingParams = getMissingParams(error.message)
    const template = getSettingMissingParamsTemplate(missingParams, params)
    setEditorContent(template)
  }

  const getMissingParams = (missingParamsErrorMessage: string) => {
    const regExp = new RegExp(`^Expected parameter\\(s\\): (.*?)$`, 'g')
    const match = regExp.exec(missingParamsErrorMessage)
    if (match && match.length > 1) {
      return match[1].split(', ')
    } else {
      return []
    }
  }
  const getSettingMissingParamsTemplate = (
    missingParams: string[],
    existingParams: Record<string, unknown>
  ): string => {
    const missingParamsTemplate = missingParams
      .map(param => `  "${param}": fill_in_your_value`)
      .join(',\n')

    let existingParamsTemplate = ''
    const existingParamsIsEmpty = Object.keys(existingParams).length === 0
    if (!existingParamsIsEmpty) {
      const existingParamsStringWithBracketsAndSurroundingNewlines =
        stringifyMod(existingParams, stringModifier, true)
      const existingParamsStringCleaned =
        existingParamsStringWithBracketsAndSurroundingNewlines
          .slice(
            1,
            existingParamsStringWithBracketsAndSurroundingNewlines.length - 1
          )
          .trim()
      existingParamsTemplate = `,\n\n  ${existingParamsStringCleaned}`
    }

    return `:params \n{\n${missingParamsTemplate}${existingParamsTemplate}\n}`
  }

  return (
    <StyledMissingParamsTemplateLink>
      <span>Use this template to add missing parameter(s):</span>
      <StyledParamsTemplateClickableArea
        onClick={() => handleTemplateHelpMessageClick()}
        aria-label={
          'Set editor content with template to be used for setting the missing parameters'
        }
      >
        :params{' {'}
        <StyledSpecifyParamsText>specify params</StyledSpecifyParamsText>
        {'}'}
      </StyledParamsTemplateClickableArea>
    </StyledMissingParamsTemplateLink>
  )
}
