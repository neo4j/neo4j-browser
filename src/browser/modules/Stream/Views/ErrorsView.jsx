/*
 * Copyright (c) 2002-2017 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { withBus } from 'preact-suber'
import { executeCommand } from 'shared/modules/commands/commandsDuck'
import { listAvailableProcedures } from 'shared/modules/cypher/procedureFactory'
import { isUnknownProcedureError } from 'services/cypherErrorsHelper'
import Visible from 'browser-components/Visible'
import { PlayIcon } from 'browser-components/icons/Icons'
import {
  StyledCypherErrorMessage,
  StyledHelpContent,
  StyledH4,
  StyledPreformattedArea,
  StyledHelpDescription,
  StyledDiv,
  StyledLink,
  StyledLinkContainer,
  StyledHelpFrame
} from '../styled'

const ErrorsView = ({error, style, bus}) => {
  if (!error) {
    return null
  }
  return (
    <StyledHelpFrame style={style}>
      <StyledHelpContent>
        <StyledHelpDescription>
          <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
          <StyledH4>{error.code}</StyledH4>
        </StyledHelpDescription>
        <StyledDiv>
          <StyledPreformattedArea>{error.message}</StyledPreformattedArea>
        </StyledDiv>
        <Visible if={isUnknownProcedureError(error)}>
          <StyledLinkContainer>
            <StyledLink onClick={() => onItemClick(bus)}><PlayIcon />&nbsp;List available procedures</StyledLink>
          </StyledLinkContainer>
        </Visible>
      </StyledHelpContent>
    </StyledHelpFrame>)
}

const onItemClick = (bus) => {
  const action = executeCommand(listAvailableProcedures)
  bus.send(action.type, action)
}

export default withBus(ErrorsView)
