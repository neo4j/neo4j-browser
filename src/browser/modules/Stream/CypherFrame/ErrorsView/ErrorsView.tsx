/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { withBus } from 'react-suber'
import { Action, Dispatch } from 'redux'
import { Bus } from 'suber'

import { PlayIcon } from 'browser-components/icons/LegacyIcons'

import { errorMessageFormater } from '../../errorMessageFormater'
import {
  StyledCypherErrorMessage,
  StyledDiv,
  StyledErrorH4,
  StyledHelpContent,
  StyledHelpDescription,
  StyledHelpFrame,
  StyledLink,
  StyledLinkContainer,
  StyledPreformattedArea
} from '../../styled'
import { MissingParamsTemplateLink } from './MissingParamsTemplateLink'
import { GlobalState } from 'project-root/src/shared/globalState'
import {
  commandSources,
  executeCommand,
  listDbsCommand
} from 'project-root/src/shared/modules/commands/commandsDuck'
import { getListProcedureQuery } from 'shared/modules/cypher/functionsAndProceduresHelper'
import * as editor from 'project-root/src/shared/modules/editor/editorDuck'
import { getParams } from 'project-root/src/shared/modules/params/paramsDuck'
import { BrowserRequestResult } from 'project-root/src/shared/modules/requests/requestsDuck'
import { GENERATE_SET_MISSING_PARAMS_TEMPLATE } from 'project-root/src/shared/modules/udc/udcDuck'
import {
  isImplicitTransactionError,
  isNoDbAccessError,
  isParameterMissingError,
  isUnknownProcedureError
} from 'services/cypherErrorsHelper'
import { BrowserError } from 'services/exceptions'
import { deepEquals } from 'neo4j-arc/common'
import { getSemanticVersion } from 'shared/modules/dbMeta/dbMetaDuck'
import { SemVer } from 'semver'

export type ErrorsViewProps = {
  result: BrowserRequestResult
  bus: Bus
  neo4jVersion: SemVer | null
  params: Record<string, unknown>
  executeCmd: (cmd: string) => void
  setEditorContent: (cmd: string) => void
}

class ErrorsViewComponent extends Component<ErrorsViewProps> {
  shouldComponentUpdate(props: ErrorsViewProps): boolean {
    return (
      !deepEquals(props.result, this.props.result) ||
      !deepEquals(props.params, this.props.params)
    )
  }

  render(): null | JSX.Element {
    const { bus, params, executeCmd, setEditorContent, neo4jVersion } =
      this.props

    const error = this.props.result as BrowserError
    if (!error || !error.code) {
      return null
    }
    const fullError = errorMessageFormater(null, error.message)

    const handleSetMissingParamsTemplateHelpMessageClick = () => {
      bus.send(GENERATE_SET_MISSING_PARAMS_TEMPLATE, undefined)
    }

    return (
      <StyledHelpFrame>
        <StyledHelpContent>
          <StyledHelpDescription>
            <StyledCypherErrorMessage>ERROR</StyledCypherErrorMessage>
            <StyledErrorH4>{error.code}</StyledErrorH4>
          </StyledHelpDescription>
          <StyledDiv>
            <StyledPreformattedArea data-testid={'cypherFrameErrorMessage'}>
              {fullError.message}
            </StyledPreformattedArea>
          </StyledDiv>
          {isUnknownProcedureError(error) && (
            <StyledLinkContainer>
              <StyledLink
                onClick={() => executeCmd(getListProcedureQuery(neo4jVersion))}
              >
                <PlayIcon />
                &nbsp;List available procedures
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isNoDbAccessError(error) && (
            <StyledLinkContainer>
              <StyledLink onClick={() => executeCmd(`:${listDbsCommand}`)}>
                <PlayIcon />
                &nbsp;List available databases
              </StyledLink>
            </StyledLinkContainer>
          )}
          {isImplicitTransactionError(error) && (
            <StyledLinkContainer>
              <StyledLink onClick={() => executeCmd(`:help auto`)}>
                <PlayIcon />
                &nbsp;Info on the <code>:auto</code> command
              </StyledLink>
              &nbsp;(auto-committing transactions)
            </StyledLinkContainer>
          )}
          {isParameterMissingError(error) && (
            <MissingParamsTemplateLink
              error={error}
              params={params}
              setEditorContent={setEditorContent}
              onTemplateHelpMessageClick={
                handleSetMissingParamsTemplateHelpMessageClick
              }
            />
          )}
        </StyledHelpContent>
      </StyledHelpFrame>
    )
  }
}

const mapStateToProps = (state: GlobalState) => {
  return {
    params: getParams(state),
    neo4jVersion: getSemanticVersion(state)
  }
}
const mapDispatchToProps = (
  _dispatch: Dispatch<Action>,
  ownProps: ErrorsViewProps
) => {
  return {
    executeCmd: (cmd: string) => {
      const action = executeCommand(cmd, {
        source: commandSources.button
      })
      ownProps.bus.send(action.type, action)
    },
    setEditorContent: (cmd: string) => {
      ownProps.bus.send(editor.SET_CONTENT, editor.setContent(cmd))
    }
  }
}
export const ErrorsView = withBus(
  connect(mapStateToProps, mapDispatchToProps)(ErrorsViewComponent)
)
