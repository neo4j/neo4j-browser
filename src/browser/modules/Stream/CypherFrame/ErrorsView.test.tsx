/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import { render } from '@testing-library/react'
import React from 'react'

import { ErrorsStatusbar, ErrorsView } from './ErrorsView'

describe('ErrorsViews', () => {
  describe('ErrorsView', () => {
    test('displays nothing if no errors', () => {
      // Given
      const props = {
        result: {}
      }

      // When
      const { container } = render(<ErrorsView {...props} />)

      // Then
      expect(container).toMatchSnapshot()
    })
    test('does displays an error', () => {
      // Given
      const props = {
        result: {
          code: 'Test.Error',
          message: 'Test error description'
        }
      }

      // When
      const { container } = render(<ErrorsView {...props} />)

      // Then
      expect(container).toMatchSnapshot()
    })
    test('displays procedure link if unknown procedure', () => {
      // Given
      const procErrorCode = 'Neo.ClientError.Procedure.ProcedureNotFound'
      const props = {
        result: {
          code: procErrorCode,
          message: 'not found'
        }
      }

      // When
      const { container, getByText } = render(<ErrorsView {...props} />)

      // Then
      expect(container).toMatchSnapshot()
      expect(getByText('List available procedures')).not.toBeUndefined()
    })
    test('displays procedure link if periodic commit error', () => {
      // Given
      const props = {
        result: {
          code: 'Neo.ClientError.Statement.SemanticError',
          message:
            'Executing queries that use periodic commit in an open transaction is not possible.'
        }
      }

      // When
      const { getByText } = render(<ErrorsView {...props} />)

      // Then
      // We need to split up because of the use of <code> tags in the rendered document
      expect(getByText(/info on the/i)).not.toBeUndefined()
      expect(getByText(':auto')).not.toBeUndefined()
      expect(getByText('(auto-committing transactions)')).not.toBeUndefined()
    })
  })
  describe('ErrorsStatusbar', () => {
    test('displays nothing if no error', () => {
      // Given
      const props = {
        result: {}
      }

      // When
      const { container } = render(<ErrorsStatusbar {...props} />)
      expect(container).toMatchSnapshot()
    })
    test('displays error', () => {
      // Given
      const props = {
        result: {
          code: 'Test.Error',
          message: 'Test error description'
        }
      }

      // When
      const { container } = render(<ErrorsStatusbar {...props} />)
      expect(container).toMatchSnapshot()
    })
  })
})
