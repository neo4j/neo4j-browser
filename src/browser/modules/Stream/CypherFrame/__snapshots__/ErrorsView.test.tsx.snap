// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorsViews ErrorsStatusbar displays error 1`] = `
<div>
  <div
    class="sc-bdVaJa kwzqrW"
  >
    <span
      class="sc-dxgOiQ eBOcVg"
      title="Test.Error: Test error description"
    >
      <i
        class="sc-fAjcbJ iasrDD fa fa-exclamation-triangle"
      />
       
      Test.Error: Test error description
    </span>
  </div>
</div>
`;

exports[`ErrorsViews ErrorsStatusbar displays nothing if no error 1`] = `<div />`;

exports[`ErrorsViews ErrorsView displays nothing if no errors 1`] = `<div />`;

exports[`ErrorsViews ErrorsView displays procedure link if unknown procedure 1`] = `
<div>
  <div
    class="sc-gzVnrw ksKCRj"
  >
    <div
      class="sc-htoDjs MMTIn"
    >
      <div
        class="sc-dnqmqq bpbKgH"
      >
        <div
          class="sc-VigVT sc-fjdhpX egUJPX"
        >
          ERROR
        </div>
        <h4
          class="sc-kgoBCf kvyvll"
        >
          Neo.ClientError.Procedure.ProcedureNotFound
        </h4>
      </div>
      <div
        class="sc-iwsKbI jQhHGA"
      >
        <pre
          class="sc-kpOJdX hkuHeU"
        >
          not found
        </pre>
      </div>
      <div
        class="sc-gqjmRU kjuOwi"
      >
        <a
          class="sc-gZMcBi jPwkqT"
        >
          <i
            class="sc-dVhcbM bfSTan fa fa-play-circle-o"
          />
          Â List available procedures
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorsViews ErrorsView does displays an error 1`] = `
<div>
  <div
    class="sc-gzVnrw ksKCRj"
  >
    <div
      class="sc-htoDjs MMTIn"
    >
      <div
        class="sc-dnqmqq bpbKgH"
      >
        <div
          class="sc-VigVT sc-fjdhpX egUJPX"
        >
          ERROR
        </div>
        <h4
          class="sc-kgoBCf kvyvll"
        >
          Test.Error
        </h4>
      </div>
      <div
        class="sc-iwsKbI jQhHGA"
      >
        <pre
          class="sc-kpOJdX hkuHeU"
        >
          Test error description
        </pre>
      </div>
    </div>
  </div>
</div>
`;
